rtos/ThisThread.h

```diff
@@ -0,0 +1,188 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2018 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef THIS_THREAD_H
+#define THIS_THREAD_H
+
+#include <stdint.h>
+#include "cmsis_os2.h"
+#include "mbed_rtos1_types.h"
+#include "mbed_rtos_storage.h"
+#include "platform/Callback.h"
+#include "platform/mbed_toolchain.h"
+#include "platform/NonCopyable.h"
+#include "rtos/Semaphore.h"
+#include "rtos/Mutex.h"
+
+namespace rtos {
+/** \addtogroup rtos */
+/** @{*/
+/**
+ * \defgroup rtos_ThisThread ThisThread namespace
+ * @{
+ */
+
+/** The ThisThread namespace allows controlling the current thread.
+ *
+ *  Example:
+ *  @code
+ *  #include "mbed.h"
+ *  #include "rtos.h"
+ *
+ *  Thread thread;
+ *  DigitalOut led1(LED1);
+ *
+ *  #define STOP_FLAG 1
+ *
+ *  // Blink function toggles the led in a long running loop
+ *  void blink(DigitalOut *led) {
+ *      while (!ThisThread::flags_wait_any_for(STOP_FLAG, 1000)) {
+ *          *led = !*led;
+ *      }
+ *  }
+ *
+ *  // Spawns a thread to run blink for 5 seconds
+ *  int main() {
+ *      thread.start(callback(blink, &led1));
+ *      ThisThread::sleep_for(5000);
+ *      thread.signal_set(STOP_FLAG);
+ *      thread.join();
+ *  }
+ *  @endcode
+ *
+ */
+namespace ThisThread {
+    /** Clears the specified Thread Flags of the currently running thread.
+      @param   flags  specifies the thread flags of the thread that should be cleared.
+      @return  thread flags before clearing.
+      @note You cannot call this function from ISR context.
+      @see Thread::flags_set
+    */
+    uint32_t flags_clear(uint32_t flags);
+
+    /** Returns the Thread Flags currently set for the currently running thread.
+      @return  current thread flags or 0 if not in a valid thread.
+      @note You may call this function from ISR context.
```

Comment:
Actually it isn't legal, according to docs.

The RTX implementation of `osThreadFlagsGet` does return 0 without error if in ISR context, but the CMSIS-RTOS 2.1.2 documentation says it can't be called from interrupt context. Will modify.